Less strict type requirements

kahaaga opened the pull request:
## Problem 

When working with points in (low-ish) N-dimensional space, it is typical in Julia to represent these points by statically sized vectors using the `SVector` type from [StaticArrays.jl](https://github.com/JuliaArrays/StaticArrays.jl). To pass these points on to `MiniQhull.delaunay`, I first have to gather the points, then convert them to a vector with column-major ordering, which is very slow.

```julia
using MiniQhull, StaticArrays
dim = 5
npts = 300
pts = [SVector{dim, Float64}(rand(dim) for i = 1:npts]

# Works, but is veery slow and allocates a lot of memory
x = Array(vcat(pts...,))
delaunay(dim, npts, x)
```

However, it is possible to `reinterpret` the points as a column-major vector with almost zero extra cost. But currently, using reinterpreted data doesn't work with `MiniQhull.delaunay` because it is hard-coded that the input points must be a `Vector{Float64}`. 

```julia
using MiniQhull, StaticArrays
dim = 5
npts = 300
pts = [SVector{dim, Float64}(rand(dim) for i = 1:npts]

# This operation is fast/allocates almost nothing
x = reinterpret(Float64, pts) #`x` is now of type `Base.ReinterpretArray`

# Fails, because the input must be `Vector{Float64}`
delaunay(dim, npts, x)

# Works, but adds an additional costly conversion step 
delaunay(dim, npts, Vector{Float64(x))
```

In my applications, I work with a large number of points. Converting between data formats then quickly becomes computationally unfeasible, so it is crucial performance-wise to be able to not explicitly convert to `Vector{Float64}` before passing the data to `delaunay`.

I think this package would be more flexible if `delaunay` could accept any subtype of `AbstractVector{Float64}` as input data.

## Solution 

This PR solves the problem described above by having less strict type requirements (i.e. using `AbstractVector`/`AbstractArray` instead of `Vector`/`Array`). The current behaviour is not affected, but it is now possible to use `Base.ReinterpretArray` (and other similar types) as input.

## Documentation

I added an extra use case to the documentation, similar to the examples above, highlighting how reinterpreted static vectors can be used as input to `delaunay`.

## Other 

This PR also includes the GitHub Actions CI testing that I added in a separate PR.

kahaaga committed:
Less strict type requirements

kahaaga committed:
Relax type requirements

kahaaga committed:
Add relevant example to readme

kahaaga committed:
No need to convert if vector if input is already AbstractVector

kahaaga committed:
Add CI

kahaaga committed:
Make sure coordinates passed on to _delaunay are Float64s

kahaaga committed:
Always use Float64

kahaaga committed:
Add extra method to convert when input data are integers

kahaaga committed:
Less strict type requirement

fverdugo reviewed (changes requested):
Hi @kahaaga, thanks for the PR!

I agree it is a good idea to accept vectors of vectors for representing the coordinates. I would do it in a slightly diferentment way than you have proposed.

now we have 2 signatures 
```julia
delaunay(dim::Integer, numpoints::Integer, points::Vector, flags::AbstractString) -> Matrix{Int32}
delaunay(points::Matrix, flags::AbstractString) -> Matrix{Int32}
```

I would add a new one for vectors of vectors (in this case you don't need to pass dim and numpoints as they an be discovered from the input with `length(eltype(points))`, `length(points)`). We can also extend previous signatures to more arbitrary types. The final signatures would be:

```julia
delaunay(dim::Integer, numpoints::Integer, points::AbstractVector, flags::AbstractString) -> Matrix{Int32}
delaunay(points::AbstractMatrix, flags::AbstractString) -> Matrix{Int32}
delaunay(points::AbstractVector, flags::AbstractString) -> Matrix{Int32}
```

kahaaga committed:
Add method for vector of vectors, and even less strict type requirements

kahaaga committed:
Add docstring to `delaunay`

kahaaga commented:
## Vector of vectors

> I agree it is a good idea to accept vectors of vectors for representing the coordinates. I would do it in a slightly diferentment way than you have proposed.

I added a third method for vectors of vectors. 

```julia
delaunay(points::AbstractVector{T}, flags::AbstractString) where T -> Matrix{Int32}
```

where T is some vector-like type. Behind the scenes, there are actually two `delaunay(points, flags::AbstractString) -> Matrix{Int32}` methods: 

- one for a vector of tuples (and other similar types, like `SVector`), which has the same memory layout as a column-major vector and can be reintepreted without extra memory allocations
- one for vector of regular vectors (which need to explicitly be converted)

Was it something like this you had in mind? 
 
## Arbitrary types 

> We can also extend previous signatures to more arbitrary types.

In addition to what was already in the PR, I also added 

```julia
delaunay(points::AbstractMatrix, flags::AbstractString) -> Matrix{Int32}
```

## Tests

I added tests for three common use cases: representing points as `Vector{Vector}`, `Vector{Tuple}`, and `Vector{SVector}`. 

Note: the dependency on `StaticArrays.jl` is only for the CI tests, and will not affect users installing the package. 

## Docstring

Finally, it would be nice to have access to the documentation directly from the REPL. I therefore added a docstring to the `delaunay` function.

kahaaga committed:
Add new signature, and update the other signatures, in readme

fverdugo commented:
Wow I am impressed! @kahaaga 

Thanks for the good job!

kahaaga commented:
> Thanks for the good job!

And thanks for making this package to begin with! Now I can use QHull in my Julia packages without depending on python! 🎆
